crates/ide/src/inlay_hints/closing_brace.rs

```diff
@@ -154,11 +190,171 @@ pub(super) fn hints(
+        (Some(pat), None) => format!("for {pat}"),
+        (None, Some(iterable)) => format!("for in {iterable}"),
+        _ => "for".into(),
+    }
+}
+
+fn format_match_label(match_expr: &ast::MatchExpr, config: &InlayHintsConfig<'_>) -> String {
+    match match_expr.expr() {
+        Some(expr) => {
+            let snippet = snippet_from_node(expr.syntax(), config);
+            if snippet.is_empty() { "match".into() } else { format!("match {snippet}") }
+        }
+        None => "match".into(),
+    }
+}
+
+fn format_match_arm_label(match_arm: &ast::MatchArm, config: &InlayHintsConfig<'_>) -> String {
+    let mut parts = Vec::new();
+    if let Some(pat) = match_arm.pat() {
+        let snippet = snippet_from_node(pat.syntax(), config);
+        if !snippet.is_empty() {
+            parts.push(snippet);
+        }
+    }
+    if let Some(guard) = match_arm.guard().and_then(|guard| guard.condition()) {
+        let snippet = snippet_from_node(guard.syntax(), config);
+        if !snippet.is_empty() {
+            parts.push(format!("if {snippet}"));
+        }
+    }
+
+    if parts.is_empty() { "match arm".into() } else { format!("{} =>", parts.join(" ")) }
+}
+
+fn block_is_match_arm_body(arm: &ast::MatchArm, block: &ast::BlockExpr) -> bool {
+    arm.expr()
+        .and_then(|expr| match expr {
+            ast::Expr::BlockExpr(expr_block) => Some(expr_block.syntax() == block.syntax()),
+            _ => None,
+        })
+        .unwrap_or(false)
+}
+
+fn label_for_if_block(
+    if_expr: &ast::IfExpr,
+    block: &ast::BlockExpr,
+    config: &InlayHintsConfig<'_>,
+) -> Option<String> {
+    if if_expr.then_branch().is_some_and(|then_branch| then_branch.syntax() == block.syntax()) {
+        Some(keyword_with_condition("if", if_expr.condition(), config))
+    } else if matches!(
+        if_expr.else_branch(),
+        Some(ast::ElseBranch::Block(else_block)) if else_block.syntax() == block.syntax()
+    ) {
+        Some("else".into())
+    } else {
+        None
+    }
+}
+
+fn format_let_else_label(let_else: &ast::LetElse, config: &InlayHintsConfig<'_>) -> String {
+    let let_stmt = let_else.syntax().parent().and_then(ast::LetStmt::cast);
+    let Some(stmt) = let_stmt else {
+        return "let ... else".into();
+    };
+
+    let pat = stmt.pat().and_then(|pat| {
+        let snippet = snippet_from_node(pat.syntax(), config);
+        (!snippet.is_empty()).then_some(snippet)
+    });
+    let initializer = stmt.initializer().and_then(|expr| {
+        let snippet = snippet_from_node(expr.syntax(), config);
+        (!snippet.is_empty()).then_some(snippet)
+    });
+
+    match (pat, initializer) {
+        (Some(pat), Some(init)) => format!("let {pat} = {init} else"),
+        (Some(pat), None) => format!("let {pat} else"),
+        _ => "let ... else".into(),
+    }
+}
+
+fn snippet_from_node(node: &SyntaxNode, config: &InlayHintsConfig<'_>) -> String {
```

Comment:
This function is too convoluted. IMO the criteria should be:

 - If the node contains a newline, don't render it at all (ellipsis always).
 - Otherwise, split by space and take words until the end or until the limit was reached.